### PR TITLE
[IOTDB-5341] Support GROUP BY VARIATION in aggregation query (front end part)

### PR DESCRIPTION
**Pull request by ycycse**
Design document: https://apache-iotdb.feishu.cn/docx/ELUMdYck2ov6V8xDJeAc95Xbnpg

### CODE REVIEW

**Review by liuminghui233 — CHANGES_REQUESTED** [review]

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4 [hunk 1]
```diff
@@ -475,6 +475,11 @@ groupByAttributeClause
     : TIME? LR_BRACKET (timeRange COMMA)? interval=DURATION_LITERAL (COMMA step=DURATION_LITERAL)? RR_BRACKET
     | LEVEL operator_eq INTEGER_LITERAL (COMMA INTEGER_LITERAL)*
     | TAGS LR_BRACKET identifier (COMMA identifier)* RR_BRACKET
+    | VARIATION LR_BRACKET expression (COMMA delta=INTEGER_LITERAL|realLiteral)? (COMMA keyValuePair)* RR_BRACKET
```
[the remark this answers is not in the document]
Comment (ycycse) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4:478]: done

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4 [hunk 2]
```diff
@@ -475,6 +475,11 @@ groupByAttributeClause
...
+    ;
+
+keyValuePair
```
[the remark this answers is not in the document]
Comment (ycycse) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlParser.g4:481]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Accumulator.java [hunk 3]
```diff
@@ -27,11 +27,16 @@
 public interface Accumulator {
 
   /**
-   * Column should be like: | ControlColumn | Time | Value |
+   * Column should be like: | ControlColumn | Time | Value | IgnoringNull is required when
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Accumulator.java:30]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Accumulator.java [hunk 4]
```diff
@@ -27,11 +27,16 @@
...
+   * considering the value with null-controlColumn
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/aggregation/Accumulator.java:31]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java [hunk 5]
```diff
@@ -180,6 +184,12 @@ public class AnalyzeVisitor extends StatementVisitor<Analysis, MPPQueryContext>
 
   private static final IoTDBConfig CONFIG = IoTDBDescriptor.getInstance().getConfig();
 
+  private static final Expression deviceExpression =
+      new TimeSeriesOperand(new MeasurementPath(new PartialPath(DEVICE, false), TSDataType.TEXT));
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:188]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java [hunk 6]
```diff
@@ -180,6 +184,12 @@ public class AnalyzeVisitor extends StatementVisitor<Analysis, MPPQueryContext>
...
+
+  private static final Expression endTimeExpression =
+      new TimeSeriesOperand(new MeasurementPath(new PartialPath(ENDTIME, false), TSDataType.INT64));
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:191]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.iotdb.db.mpp.execution.operator.window.WindowType;
+import org.apache.iotdb.db.mpp.plan.expression.Expression;
+import org.apache.iotdb.db.mpp.plan.statement.StatementNode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class GroupByComponent extends StatementNode {
+
+  private Expression ControlColumnExpression;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+package org.apache.iotdb.db.mpp.plan.statement.component;
+
+import org.apache.iotdb.db.mpp.execution.operator.window.WindowType;
+import org.apache.iotdb.db.mpp.plan.expression.Expression;
+import org.apache.iotdb.db.mpp.plan.statement.StatementNode;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class GroupByComponent extends StatementNode {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+  private Expression ControlColumnExpression;
+  private double delta = 0;
+
+  Map<String, String> pair = null;
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByComponent.java]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+import org.apache.iotdb.db.mpp.execution.operator.window.WindowType;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+/** The parameter of `GROUP BY`. including: `GROUP BY VARIATION` */
+public class GroupByParameter {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java]: done

**Review by Wei-hao-Li — APPROVED** [review]
Have known the change about Accumulator

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java [hunk 11]
```diff
@@ -1114,7 +1139,89 @@ private TSDataType analyzeExpression(Analysis analysis, Expression expression) {
...
+      if (expressions.size() != 1) {
+        throw new SemanticException("Expression in group by should indicate one value");
+      }
+      // Aggregation expression shouldn't exist in group by clause.
+      List<Expression> aggregationExpression =
+          ExpressionAnalyzer.searchAggregationExpressions(expressions.get(0));
+      if (aggregationExpression != null && aggregationExpression.size() != 0) {
+        throw new SemanticException("Aggregation expression shouldn't exist in group by clause");
+      }
+      analyzeExpression(analysis, expressions.get(0));
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:1211]: need check data type of groupByExpression

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java [hunk 12]
```diff
@@ -1114,7 +1139,89 @@ private TSDataType analyzeExpression(Analysis analysis, Expression expression) {
...
+        }
+        Expression groupByExpressionOfOneDevice = groupByExpressionsOfOneDevice.get(0);
+
+        // Aggregation expression shouldn't exist in group by clause.
+        List<Expression> aggregationExpression =
+            ExpressionAnalyzer.searchAggregationExpressions(groupByExpressionOfOneDevice);
+        if (aggregationExpression != null && aggregationExpression.size() != 0) {
+          throw new SemanticException("Aggregation expression shouldn't exist in group by clause");
+        }
+        analyzeExpression(analysis, groupByExpressionOfOneDevice);
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/AnalyzeVisitor.java:1171]: need check data type of groupByExpression

**Review by liuminghui233 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/RowBasedTimeJoinOperator.java [hunk 13]
```diff
@@ -210,6 +214,7 @@ public boolean hasNext() {
       return false;
     }
     if (retainedTsBlock != null) {
+      LOGGER.info("Invalid true");
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/join/RowBasedTimeJoinOperator.java:217]: removed

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByVariationParameter.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+public class GroupByVariationParameter extends GroupByParameter {
+
+  double delta;
+
+  public GroupByVariationParameter(WindowType windowType, boolean ignoringNull, double delta) {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByVariationParameter.java:34]: the parameter is received as GroupByParameter. So we need windowType to judge what kind of parameter it is.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByVariationComponent.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.apache.iotdb.db.mpp.plan.statement.component;
+
+import org.apache.iotdb.db.mpp.execution.operator.window.WindowType;
+
+public class GroupByVariationComponent extends GroupByComponent {
+  private double delta = 0;
+
+  public GroupByVariationComponent(WindowType windowType) {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByVariationComponent.java:27]: the same.

File: integration-test/src/test/java/org/apache/iotdb/db/it/groupby/IoTDBGroupByVariationIT.java [hunk 16]
```diff
@@ -0,0 +1,492 @@
...
+        "INSERT INTO root.ln.wf01.wt02(timestamp, temperature, status, hardware) values(620, 39.2, true, 550)",
+        "flush",
+        "INSERT INTO root.ln.wf01.wt02(timestamp, temperature, status, hardware) values(1500, 9.8, false, 666)",
+        "INSERT INTO root.ln.wf01.wt02(timestamp, temperature, status, hardware) values(1550, 10.2, true, 888)",
+        "flush"
+      };
+
+  @BeforeClass
+  public static void setUp() throws Exception {
+    EnvFactory.getEnv().getConfig().getCommonConfig().setPartitionInterval(1000);
```
[the remark this answers is not in the document]
Comment (ycycse) [reply integration-test/src/test/java/org/apache/iotdb/db/it/groupby/IoTDBGroupByVariationIT.java:144]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    serializeAttributes(buffer);
+  }
+
+  public void serialize(DataOutputStream stream) throws IOException {
+    ReadWriteIOUtils.write(windowType.getType(), stream);
+    ReadWriteIOUtils.write(ignoringNull, stream);
+    serializeAttributes(stream);
+  }
+
+  public boolean equals(Object obj) {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java:67]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    if (!(obj instanceof GroupByParameter)) {
+      return false;
+    }
+    GroupByParameter other = (GroupByParameter) obj;
+    return this.windowType == other.windowType && this.ignoringNull == other.ignoringNull;
+  }
+
+  public int hashCode() {
```
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByParameter.java:75]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByVariationParameter.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/planner/plan/parameter/GroupByVariationParameter.java:34]: done

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByVariationComponent.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ycycse) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/statement/component/GroupByVariationComponent.java:27]: done

**Review by liuminghui233 — APPROVED** [review]
LGTM :-)

**Pull request merged**